internal/dataplane/parser/translate_vars.go

```diff
@@ -17,4 +23,22 @@ const (
 	// DefaultHTTPPort is the network port that should be assumed by default
 	// for HTTP traffic to services.
 	DefaultHTTPPort = 80
+
+	// kongHeaderRegexPrefix is a reserved prefix string that Kong uses to determine if it should parse a header value
+	// as a regex.
+	kongHeaderRegexPrefix = "~*"
+
+	// kongPathRegexPrefix is the reserved prefix string that instructs Kong 3.0+ to interpret a path as a regex.
+	kongPathRegexPrefix = "~"
+)
+
+var (
+	// MinRegexHeaderKongVersion is the minimum Kong version that supports regex header matches.
+	MinRegexHeaderKongVersion = semver.MustParse("2.8.0")
+
+	// MinExplicitPathRegexKongVersion is the minimum Kong version that requires explicit indication of regex paths.
+	MinExplicitPathRegexKongVersion = semver.MustParse("3.0.0")
+
+	// LegacyRegexPathExpression is the regular expression used by Kong <3.0 to determine if a path is a regex
+	LegacyRegexPathExpression = regexp.MustCompile(`^\^?[a-zA-Z0-9\.\-_~/%]*\$?$`)
```

Comment:
This could not match `/regex-\d{3}-httpbin`, which appeared in `TestIngressEssentials`.

Reply:
We may not want to fix this, opting to instead change the test pattern. Actually changing the heuristic is iffy since we're not really committing to supporting it insofar as we say it actually works. It's a compatibility shim to allow you to upgrade without rewriting configuration all at once, so I'd err on the side of sticking to the known flawed behavior in Kong 2.x.

We don't want to risk any regression from what Kong 2.x did support, and the safest option for that is using the exact regex it did: regexes that were correctly detected remain so, regexes that were not remain broken. If you have a regex that's broken to be not broken, you have to migrate to the new format. You can do so piecemeal, and it's now opt-in whether the heuristic runs at all, so I'm more comfortable with the brokenness than I would be otherwise.

Reply:
Discussed with Sean, decided to go with the known, but broken one.